### PR TITLE
SystemEvents in Heimdal

### PR DESCRIPTION
**Pull request by leogrim**
@stkem 
Merging as it's working, will be glad to discuss reflection features later :)

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/heimdal/Event.scala [hunk 1]
```diff
@@ -106,15 +105,49 @@ class UserEventContextBuilderFactory @Inject() (serviceDiscovery: ServiceDiscove
...
+      val repo = (json \ "repo").as[String]
+      val event = (json \ "event")
+      repo match {
+        case Event.system => SystemEvent.format.reads(event)
+        case Event.user => UserEvent.format.reads(event)
+      }
+    }
+  }
+}
+
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/heimdal/Event.scala]: @stkem see this

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventTrackingController.scala [hunk 2]
```diff
@@ -2,21 +2,24 @@ package com.keepit.heimdal.controllers
...
 
   private[controllers] def trackInternalEvent(eventJs: JsValue) = {
-    val event: UserEvent = eventJs.as[UserEvent]
-    userEventLoggingRepo.insert(event)
+    val event: Event = eventJs.as[Event]
+    event match {
+      case systemEvent: SystemEvent => systemEventLoggingRepo.insert(systemEvent)
+      case userEvent: UserEvent => userEventLoggingRepo.insert(userEvent)
+    }
   }
```
Comment (leogrim) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventTrackingController.scala:22]: @stkem and this

**Comment by leogrim** [comment]
@stkem @andrewconner  
This is working, and introduces some candies for polymorphic / type-safe serialization (though I think I could have managed without it, but it's fun Scala and we can likely reuse it to deal with any kind of untyped external storage, e.g. a graph database). Any comment?

**Comment by stkem** [comment]
Perhaps slightly on the arcane side, but I like it! Maybe you can talk a bit about additional use cases on monday.

**Comment by andrewconner** [comment]
Mind walking me through `Companion` tomorrow? Want to make sure I understand it and its use cases.

**Comment by leogrim** [comment]
Sure, in short it's a pattern to easily build custom reflexion tools. Right
now, it's used to handle polymorphic serialization / deserialization in a
transparent way. Going forward, it looks like a potential approach to
enforce typed external ids. Not sure what else we could think of, but it is
quite flexible. It enables us to reason over companions objects and get us
access to data / functions attached to a type, even when the actual type is
unknown and an instance is not available.

2013/11/10 Andrew Conner notifications@github.com

**Pull request merged**